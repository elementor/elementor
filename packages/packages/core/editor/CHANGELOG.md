# Change Log

## 0.21.0

### Minor Changes

- 68ce0d9: Ensure current user exist in the editor.
- 1a37b1c: Update the Elementor UI version.

### Patch Changes

- Updated dependencies [68ce0d9]
  - @elementor/editor-current-user@0.6.0

## 0.20.1

### Patch Changes

- @elementor/editor-v1-adapters@0.12.1

## 0.20.0

### Minor Changes

- 17b73ab: Update `@elementor/ui` version.

## 0.19.4

### Patch Changes

- 615baa1: upgrade elementor/ui version

## 0.19.3

### Patch Changes

- Updated dependencies [7439d73]
  - @elementor/editor-v1-adapters@0.12.0

## 0.19.2

### Patch Changes

- Updated dependencies [d05046f]
  - @elementor/locations@0.8.0

## 0.19.1

### Patch Changes

- ca016cc: Elementor ui update to version 1.34.2, elementor icons update to 1.40.1

## 0.19.0

### Minor Changes

- 10cbbe9: update `@elementor/ui` version

## 0.18.6

### Patch Changes

- 42c42ee: Remove side-effects

## 0.18.5

### Patch Changes

- 666ffdd: Change theme provider styles for editing panel and class manager panel

## 0.18.4

### Patch Changes

- 3e108d9: update elementor/ui
- Updated dependencies [3e108d9]
  - @elementor/locations@0.7.7

## 0.18.3

### Patch Changes

- 0909b4b: update the toggle button group to be responsive
- Updated dependencies [e798985]
- Updated dependencies [4c2935b]
  - @elementor/editor-v1-adapters@0.11.0

## 0.18.2

### Patch Changes

- Updated dependencies [959e02c]
  - @elementor/editor-v1-adapters@0.10.2

## 0.18.1

### Patch Changes

- @elementor/editor-v1-adapters@0.10.1

## 0.18.0

### Minor Changes

- 910044c: update `@elementor/icons` and `@elementor/ui` packages.
- 5caf78d: update `@elementor/ui` package.

### Patch Changes

- Updated dependencies [da38fa9]
- Updated dependencies [ad6fde0]
- Updated dependencies [b8e2a85]
  - @elementor/editor-v1-adapters@0.10.0

## 0.17.5

### Patch Changes

- 6c4d4a7: refactored CSS class menu and fix it's keyboard navigation
- Updated dependencies [a8b60c9]
- Updated dependencies [1a9c943]
  - @elementor/editor-v1-adapters@0.9.1
  - @elementor/store@0.9.0

## 0.17.4

### Patch Changes

- cfbd198: Update `@elementor/ui` version

## 0.17.3

### Patch Changes

- Updated dependencies [65d3c4c]
  - @elementor/editor-v1-adapters@0.9.0

## 0.17.2

### Patch Changes

- 91453b3: Update and lock dependencies versions
- Updated dependencies [91453b3]
  - @elementor/editor-v1-adapters@0.8.5

## 0.17.1

### Patch Changes

- 1926fe1: Update dependencies
- Updated dependencies [1926fe1]
  - @elementor/query@0.2.4

## 0.17.0

### Minor Changes

- e4c6e3b: update @elementor/ui version
  fix color picker position in box shadow repeater control
  make the color control full width

### Patch Changes

- 7781969: Update `@elementor/ui` version
- Updated dependencies [0c6bcb6]
  - @elementor/editor-v1-adapters@0.8.4

## 0.16.0

### Minor Changes

- 22f8240: Replaced `UnstableColorPicker` with `UnstableColorField` in color control

### Patch Changes

- 9abdfaf: Fix package.json exports field
- Updated dependencies [9abdfaf]
  - @elementor/editor-v1-adapters@0.8.3
  - @elementor/locations@0.7.6
  - @elementor/query@0.2.3
  - @elementor/store@0.8.6

## 0.15.0

### Minor Changes

- 701fc35: Updated color picker version

## 0.14.0

### Minor Changes

- 0e70721: Update `@elementor/ui` and `@elementor/icons` versions

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.13.1](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.13.0...@elementor/editor@0.13.1) (2024-08-20)

**Note:** Version bump only for package @elementor/editor

# [0.13.0](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.12.0...@elementor/editor@0.13.0) (2024-08-14)

### Features

- **editor-editing-panel:** init attachment control [EDS-320] ([#239](https://github.com/elementor/elementor-packages/issues/239)) ([9871e10](https://github.com/elementor/elementor-packages/commit/9871e105c98b8c4502289cf8cf99d75b4074e0a9))

# [0.12.0](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.11.2...@elementor/editor@0.12.0) (2024-08-05)

### Bug Fixes

- publish only necessary files to npm ([#226](https://github.com/elementor/elementor-packages/issues/226)) ([d808e2f](https://github.com/elementor/elementor-packages/commit/d808e2f60eb7ca2d7b8560d0b79c0e62c2f969a8))

### Features

- **editor-editing-panel:** change textarea component [EDS-311] ([#221](https://github.com/elementor/elementor-packages/issues/221)) ([a9ab0d5](https://github.com/elementor/elementor-packages/commit/a9ab0d53e87086e323c4f024dca0eae93005e1a1))
- **editor-editing-panel:** settings and style tabs [EDS-277] ([#217](https://github.com/elementor/elementor-packages/issues/217)) ([c623797](https://github.com/elementor/elementor-packages/commit/c6237974d9940a2b58bdd1df3d835dc828e323f4))

## [0.11.2](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.11.1...@elementor/editor@0.11.2) (2024-07-16)

**Note:** Version bump only for package @elementor/editor

## [0.11.1](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.11.0...@elementor/editor@0.11.1) (2024-07-16)

### Bug Fixes

- **editor:** remove extensions dependency [EDS-263] ([#204](https://github.com/elementor/elementor-packages/issues/204)) ([22ca6a4](https://github.com/elementor/elementor-packages/commit/22ca6a423818c479ef7fafc177bba6e45dd697ff))

# [0.11.0](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.10.13...@elementor/editor@0.11.0) (2024-07-08)

### Features

- **editor-editing-panel:** create editor panel package [EDS-226] ([#188](https://github.com/elementor/elementor-packages/issues/188)) ([e361aed](https://github.com/elementor/elementor-packages/commit/e361aed023c3a5d4dd329b354f1403de238da20e))

## [0.10.13](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.10.12...@elementor/editor@0.10.13) (2024-07-03)

**Note:** Version bump only for package @elementor/editor

## [0.10.12](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.10.11...@elementor/editor@0.10.12) (2024-07-02)

**Note:** Version bump only for package @elementor/editor

## [0.10.11](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.10.10...@elementor/editor@0.10.11) (2024-05-09)

**Note:** Version bump only for package @elementor/editor

## [0.10.10](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.10.9...@elementor/editor@0.10.10) (2024-05-07)

**Note:** Version bump only for package @elementor/editor

## [0.10.9](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.10.8...@elementor/editor@0.10.9) (2024-04-16)

**Note:** Version bump only for package @elementor/editor

## [0.10.8](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.10.7...@elementor/editor@0.10.8) (2024-02-28)

**Note:** Version bump only for package @elementor/editor

## [0.10.7](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.10.6...@elementor/editor@0.10.7) (2024-02-22)

**Note:** Version bump only for package @elementor/editor

## [0.10.6](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.10.5...@elementor/editor@0.10.6) (2024-01-29)

**Note:** Version bump only for package @elementor/editor

## [0.10.5](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.10.4...@elementor/editor@0.10.5) (2023-11-07)

**Note:** Version bump only for package @elementor/editor

## [0.10.4](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.10.3...@elementor/editor@0.10.4) (2023-11-02)

**Note:** Version bump only for package @elementor/editor

## [0.10.3](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.10.2...@elementor/editor@0.10.3) (2023-10-19)

**Note:** Version bump only for package @elementor/editor

## [0.10.2](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.10.1...@elementor/editor@0.10.2) (2023-10-19)

**Note:** Version bump only for package @elementor/editor

## [0.10.1](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.10.0...@elementor/editor@0.10.1) (2023-09-26)

**Note:** Version bump only for package @elementor/editor

# [0.10.0](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.9.0...@elementor/editor@0.10.0) (2023-09-14)

### Features

- **v1-adapters:** mark private functions [ED-12029] ([#122](https://github.com/elementor/elementor-packages/issues/122)) ([2e79ccc](https://github.com/elementor/elementor-packages/commit/2e79ccc87add5f0508b5a142f9f6d832b4657062))

# [0.9.0](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.8.1...@elementor/editor@0.9.0) (2023-09-11)

### Features

- **editor-documents:** mark experimental functions [ED-12066] ([#112](https://github.com/elementor/elementor-packages/issues/112)) ([f4b7a9a](https://github.com/elementor/elementor-packages/commit/f4b7a9ac7953a2b48062349495acff05b5640de3))
- **store:** mark experimental functions [ED-12070] ([#114](https://github.com/elementor/elementor-packages/issues/114)) ([d790829](https://github.com/elementor/elementor-packages/commit/d79082911195c75f6d3a89a5619b234e2f5ce158))

## [0.8.1](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.8.0...@elementor/editor@0.8.1) (2023-08-02)

**Note:** Version bump only for package @elementor/editor

# [0.8.0](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.7.3...@elementor/editor@0.8.0) (2023-08-01)

### Features

- **app-bar:** introduce integrations menu [ED-11437] ([#93](https://github.com/elementor/elementor-packages/issues/93)) ([507f367](https://github.com/elementor/elementor-packages/commit/507f367ea54e80ed29c22b64fe93f15353327c21))

## [0.7.3](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.7.2...@elementor/editor@0.7.3) (2023-07-17)

**Note:** Version bump only for package @elementor/editor

## [0.7.2](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.7.1...@elementor/editor@0.7.2) (2023-07-17)

**Note:** Version bump only for package @elementor/editor

## [0.7.1](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.7.0...@elementor/editor@0.7.1) (2023-06-29)

**Note:** Version bump only for package @elementor/editor

# [0.7.0](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.6.2...@elementor/editor@0.7.0) (2023-06-28)

### Features

- **query:** add support for React Query [ED-11193] ([#62](https://github.com/elementor/elementor-packages/issues/62)) ([d50db2f](https://github.com/elementor/elementor-packages/commit/d50db2f4b310e644ddc154620bef1a850764d9a2))

## [0.6.2](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.6.1...@elementor/editor@0.6.2) (2023-06-25)

### Bug Fixes

- **editor:** document query parameters are not synced properly [ED-11024] ([#54](https://github.com/elementor/elementor-packages/issues/54)) ([4c17080](https://github.com/elementor/elementor-packages/commit/4c17080a7cecec7c807954e9381fb7c840e51ea4))

## [0.6.1](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.6.0...@elementor/editor@0.6.1) (2023-06-11)

**Note:** Version bump only for package @elementor/editor

# [0.6.0](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.5.1...@elementor/editor@0.6.0) (2023-06-06)

### Features

- drop support for React 17 [ED-10982] ([#50](https://github.com/elementor/elementor-packages/issues/50)) ([59c576c](https://github.com/elementor/elementor-packages/commit/59c576ca218947dc0992616311d4d399a20e91a6))

## [0.5.1](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.5.0...@elementor/editor@0.5.1) (2023-06-05)

**Note:** Version bump only for package @elementor/editor

# [0.5.0](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.4.0...@elementor/editor@0.5.0) (2023-06-01)

### Features

- add support for React 18 [ED-10905] ([#46](https://github.com/elementor/elementor-packages/issues/46)) ([c8915c6](https://github.com/elementor/elementor-packages/commit/c8915c6ea62550bcdeb5c8a576f311bedc35bcad))

# [0.4.0](https://github.com/elementor/elementor-packages/compare/@elementor/editor@0.3.0...@elementor/editor@0.4.0) (2023-05-30)

### Features

- **editor-documents:** sync document name with V1 [ED-9930] ([#45](https://github.com/elementor/elementor-packages/issues/45)) ([b4d7250](https://github.com/elementor/elementor-packages/commit/b4d7250fb5ed03e21dafcf76a82756cf086ce212))

# 0.3.0 (2023-05-21)

### Features

- **editor-site-navigation:** change post without refresh [ED-10713] ([#39](https://github.com/elementor/elementor-packages/issues/39)) ([b72e82f](https://github.com/elementor/elementor-packages/commit/b72e82f9adb9c1237300cbf603f33f268f9f0400))
- **locations:** change api to support props for slots [ED-10730] ([#38](https://github.com/elementor/elementor-packages/issues/38)) ([44bec3c](https://github.com/elementor/elementor-packages/commit/44bec3cda1020037ba7105c6f05ce4baa8e3b376))

## 0.2.1 (2023-05-11)

### Bug Fixes

- **editor-app-bar:** make save options button disabled in site-settings [ED-10221] ([#30](https://github.com/elementor/elementor-packages/issues/30)) ([1991096](https://github.com/elementor/elementor-packages/commit/1991096115efeae7bc3648e4889899b85d7328d6))

# 0.2.0 (2023-05-09)

### Features

- **env:** support consuming environment variables in packages [ED-10161] ([#26](https://github.com/elementor/elementor-packages/issues/26)) ([dd811ff](https://github.com/elementor/elementor-packages/commit/dd811ff71e43bf998ee07c6e1f380a379874a7fd))

## [0.2.1](https://github.com/elementor/elementor-packages/compare/v0.2.0...v0.2.1) (2023-05-11)

### Bug Fixes

- **editor-app-bar:** make save options button disabled in site-settings [ED-10221] ([#30](https://github.com/elementor/elementor-packages/issues/30)) ([1991096](https://github.com/elementor/elementor-packages/commit/1991096115efeae7bc3648e4889899b85d7328d6))
