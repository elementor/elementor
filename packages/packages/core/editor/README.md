# Editor

> [!WARNING]
> Please refrain from accessing or depending on functions and variables starting with double underscores, as they are subject to change without notice.
> Naming convention involving double underscores (`__`) as a prefix to indicate that a function or variable is meant for internal use and should not be accessed or relied upon by third-party developers.
