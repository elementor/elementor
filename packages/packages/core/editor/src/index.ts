export { injectIntoTop, injectIntoLogic } from './locations';
export { start } from './start';
