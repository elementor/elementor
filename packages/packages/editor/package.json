{
	"name": "@elementor/editor",
	"dependencies": {
		"@elementor/locations": "*",
		"@elementor/ui": "*",
		"@elementor/store": "*",
		"@elementor/v1-adapters": "*",
		"@wordpress/i18n": "*"
	},
	"peerDependencies": {
		"react": "17.x",
		"react-dom": "17.x"
	}
}
