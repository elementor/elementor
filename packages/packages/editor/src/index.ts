export { injectIntoTop } from './locations';
export { default as init } from './init';
export { useSettings } from './contexts/settings-context';
export type { Settings } from './contexts/settings-context';
