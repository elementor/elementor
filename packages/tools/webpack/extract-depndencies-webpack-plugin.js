// Inspired by "Dependency Extraction Webpack Plugin" by @wordpress team.
// Link: https://github.com/WordPress/gutenberg/tree/trunk/packages/dependency-extraction-webpack-plugin

const { sources: { RawSource } } = require( 'webpack' );

module.exports = class ExtractDependenciesWebpackPlugin {
	generateHandleName;
	replaceDependencyNames;
	generateAssetsFileName;

	constructor( {
		generateHandleName,
		replaceDependencyNames,
		generateAssetsFileName,
	} = {} ) {
		this.generateHandleName = generateHandleName || this.defaultGenerateHandleName;
		this.replaceDependencyNames = replaceDependencyNames || this.defaultReplaceDependencyNames;
		this.generateAssetsFileName = generateAssetsFileName || this.defaultGenerateAssetsFileName;
	}

	apply( compiler ) {
		compiler.hooks.thisCompilation.tap( this.constructor.name, ( compilation ) => {
			compilation.hooks.processAssets.tap( { name: this.constructor.name }, () => {
				[ ...compilation.entrypoints ].forEach( ( [ id, entrypoint ] ) => {
					const chunk = entrypoint.chunks.find( ( { name } ) => name === id );
					const chunkJSFile = this.getFileFromChunk( chunk );

					if ( ! chunkJSFile ) {
						return;
					}

					const deps = this.getDepsFromChunk( compilation, chunk );

					const assetFilename = this.generateAssetsFileName(
						compilation.getPath( '[file]', { filename: chunkJSFile } )
					);

					const content = this.createAssetsFileContent(
						id,
						deps,
						compilation.options.optimization.minimize
					);

					// Add source and file into compilation for webpack to output.
					compilation.assets[ assetFilename ] = new RawSource( content );

					chunk.files.add( assetFilename );
				} );
			} );
		} );
	}

	getDepsFromChunk( compilation, chunk ) {
		const externals = Object.keys( compilation.options.externals );
		const depsSet = new Set();

		compilation.chunkGraph.getChunkModules( chunk ).forEach( ( module ) => {
			[ ...( module.modules || [] ), module ].forEach( ( subModule ) => {
				const isExternalDep = externals.includes( subModule.userRequest );

				if ( ! isExternalDep ) {
					return;
				}

				depsSet.add( subModule.userRequest );
			} );
		} );

		return depsSet;
	}

	createAssetsFileContent( entryId, deps, shouldMinify ) {
		const handleName = this.generateHandleName( entryId );

		const depsAsString = [ ...deps ]
			.map( ( dep ) => this.replaceDependencyNames( dep ) )
			.filter( ( dep ) => dep !== handleName )
			.sort()
			.map( ( dep ) => `'${ dep }',` )
			.join( '\n\t\t' );

		let content =
`<?php
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

return [
	'handle' => '${ handleName }',
	'deps' => [
		${ depsAsString }
	],
];
`;

		if ( shouldMinify ) {
			content = content
				.replaceAll( /\s/gm, '' )
				.replace( '<?php', '<?php\n' );
		}

		return content;
	}

	getFileFromChunk( chunk ) {
		return [ ...chunk.files ].find( ( f ) => /\.js$/i.test( f ) );
	}

	defaultReplaceDependencyNames( name ) {
		const map = new Map( [
			[ '@elementor/', 'elementor-packages-' ],
			[ '@wordpress/', 'wp-' ],
		] );

		for ( const [ key, value ] of map ) {
			if ( name.startsWith( key ) ) {
				return name.replace( key, value );
			}
		}

		return name;
	}

	defaultGenerateAssetsFileName( filename ) {
		return filename.replace( /(\.min)?\.js$/i, '.asset.php' );
	}

	defaultGenerateHandleName( name ) {
		return `elementor-packages-${ name }`;
	}
};
