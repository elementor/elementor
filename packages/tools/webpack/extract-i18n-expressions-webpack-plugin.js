const fs = require( 'fs' );
const path = require( 'path' );

const MODULE_FILTERS = Object.freeze( [ /(([^!?\s]+?)(?:\.js|\.jsx|\.ts|\.tsx))$/, /^((?!node_modules).)*$/ ] );

const COMMENTS_REGEXPS = Object.freeze( [
	// Matches translators comment block: `/* translators: %s */`.
	/\/\*[\t ]*translators:.*\*\//gm,
	/// Matches translators inline comment: `// translators: %s`.
	/(\/\/)[\t ]*translators:[^\r\n]*/gm,
] );

const TRANSLATIONS_REGEXPS = Object.freeze( [
	// Matches translation functions: `__('Hello', 'elementor')`, `_n('Me', 'Us', 2, 'elementor-pro')`.
	/\b_(?:_|n|nx|x)\(.*?,\s*(?<c>['"`])[\w-]+\k<c>\)/gm,
] );

module.exports = class ExtractI18nExpressionsWebpackPlugin {
	apply( compiler ) {
		// Learn more about Webpack plugin system: https://webpack.js.org/api/plugins/

		let translationCallExpressions = [];

		// Learn more about Webpack compilation process and hooks: https://webpack.js.org/api/compilation-hooks/
		compiler.hooks.thisCompilation.tap( this.constructor.name, ( compilation ) => {
			// We tap into the time that Webpack has finished processing all the other assets
			// learn more: https://webpack.js.org/api/compilation-hooks/#processassets.
			compilation.hooks.processAssets.tap( { name: this.constructor.name }, () => {
				translationCallExpressions = this.getModuleExpressionsMap( compilation );
			} );
		} );

		compiler.hooks.afterEmit.tapPromise( this.constructor.name, async ( compilation ) => {
			// Create all the translations files based on the call expressions.
			await this.createTranslationsFiles( compilation, translationCallExpressions );
		} );
	}

	getModuleExpressionsMap( compilation ) {
		const moduleExpressionsMap = new Map();

		[ ...compilation.chunks ].forEach( ( chunk ) => {
			const chunkJSFile = this.getFileFromChunk( chunk );

			if ( ! chunkJSFile ) {
				// There's no JS file in this chunk, no work for us. Typically a `style.css` from cache group.
				return;
			}

			compilation.chunkGraph.getChunkModules( chunk ).forEach( ( module ) => {
				this.getSubModulesToCheck( module ).forEach( ( subModule ) => {
					const mainEntryFile = this.findMainModuleOfEntry( subModule, compilation );

					if ( ! moduleExpressionsMap.has( mainEntryFile ) ) {
						moduleExpressionsMap.set( mainEntryFile, [] );
					}

					// Running over the submodules and find all the translation call expressions and their translators comment
					// (e.g `/* translators: %s: name*/ __('Hello %s', 'elementor')`),
					// extract them and add them to a Map, where the key is the main entry file, and the value is an array of all the
					// translation call expressions.
					this.extractExpressionsFromSubmodule( subModule ).forEach( ( expression ) => {
						moduleExpressionsMap.get( mainEntryFile ).push( expression );
					} );
				} );
			} );
		} );

		return moduleExpressionsMap;
	}

	async createTranslationsFiles( compilation, translationCallExpressions ) {
		const promises = [ ...compilation.entrypoints ].map( ( [ id, entrypoint ] ) => {
			const chunk = entrypoint.chunks.find( ( { name } ) => name === id );

			const chunkJSFile = this.getFileFromChunk( chunk );

			if ( ! chunkJSFile ) {
				return Promise.resolve();
			}

			const mainFilePath = compilation.options.entry[ id ].import[ 0 ];

			const assetFilename = this.generateTranslationFilename(
				compilation.options.output.path,
				compilation.getPath( '[file]', { filename: chunkJSFile } )
			);

			const assetFileContent = this.generateTranslationFileContent(
				translationCallExpressions.get( mainFilePath ) || []
			);

			return fs.promises.writeFile( assetFilename, assetFileContent );
		} );

		return await Promise.all( promises );
	}

	getSubModulesToCheck( module ) {
		return [ ...( module.modules || [] ), module ]
			.filter( ( subModule ) => this.shouldCheckModule( subModule ) );
	}

	getFileFromChunk( chunk ) {
		return [ ...chunk.files ].find( ( f ) => /\.js$/i.test( f ) );
	}

	shouldCheckModule( module ) {
		return MODULE_FILTERS.every( ( filter ) => !! filter.test( module.userRequest ) );
	}

	findMainModuleOfEntry( module, compilation ) {
		if ( compilation.moduleGraph.getIssuer( module ) ) {
			return this.findMainModuleOfEntry( compilation.moduleGraph.getIssuer( module ), compilation );
		}

		return module.rawRequest;
	}

	extractExpressionsFromSubmodule( subModule ) {
		const source = subModule?._source?._valueAsString;

		if ( ! source ) {
			return [];
		}

		const translationCallExpressions = [];

		[
			...TRANSLATIONS_REGEXPS,
			...COMMENTS_REGEXPS,
		].forEach( ( regexp ) => {
			[ ...source.matchAll( regexp ) ].forEach( ( res ) => {
				translationCallExpressions.push( {
					type: COMMENTS_REGEXPS.includes( regexp ) ? 'comment' : 'call-expression',
					index: res.index,
					value: res[ 0 ],
				} );
			} );
		} );

		return translationCallExpressions;
	}

	generateTranslationFilename( basePath, filename ) {
		return path.join(
			basePath,
			filename.replace( /(\.min)?\.js$/i, '.strings.js' )
		);
	}

	generateTranslationFileContent( expressions ) {
		return expressions
			// Sort by the index it was found in the file based on the regexp (and not by the order it was added to the array).
			.sort( ( a, b ) => a.index - b.index )
			// Add a semicolon when needed.
			.map( ( expr ) => `${ expr.value }${ expr.type === 'comment' ? '' : ';' }` )
			// Join all the expressions to a single string with line-breaks between them.
			.join( '\n' );
	}
};
