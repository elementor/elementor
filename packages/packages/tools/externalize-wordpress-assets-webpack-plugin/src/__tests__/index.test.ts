/**
 * @jest-environment node
 */

import * as fs from 'fs';
import { vol } from 'memfs';
import * as path from 'path';
import { webpack } from 'webpack';

import { ExternalizeWordPressAssetsWebpackPlugin } from '../index';

jest.mock( 'fs', () => jest.requireActual( 'memfs' ) );

describe( '@elementor/externalize-wordpress-assets-webpack-plugin', () => {
	beforeEach( () => {
		const fileContent = `
			import editor from '@elementor/editor';
			import editorPanels from '@elementor/editor-panels';
			import wp from '@wordpress/element';
			import React from 'react';
			import ReactDOM from 'react-dom';
			import other from '@other/package-name';

			editor();
			editorPanels();
			wp();
			React();
			ReactDOM();
			other();
			
			export const test = 'This is test for exporting something';
		`;

		fs.writeFileSync( path.resolve( '/app.js' ), fileContent );
	} );

	afterEach( () => {
		vol.reset();
	} );

	it( 'should externalize Elementor & WordPress assets', ( done ) => {
		// Arrange.
		const compiler = webpack( {
			mode: 'development',
			devtool: false,
			entry: { 'my-app-test': path.resolve( '/app.js' ) },
			context: path.resolve( '/app.js' ),
			optimization: { runtimeChunk: 'single' },
			output: {
				filename: '[name].js',
				path: path.resolve( '/dist' ),
			},
			plugins: [
				new ExternalizeWordPressAssetsWebpackPlugin( {
					global: ( entryName ) => [ 'elementorPackages', entryName ],
					map: [
						{ request: 'react', global: 'React' },
						{ request: 'react-dom', global: 'ReactDOM' },
						{ request: /^@elementor\/(.+)$/, global: [ 'elementorPackages', '$1' ] },
						{ request: /^@wordpress\/(.+)$/, global: [ 'wp', '$1' ] },
						{ request: '@other/package-name', global: 'otherPackageName' },
					],
				} ),
			],
		} );

		// Expect.
		expect.assertions( 4 );

		// Act.
		compiler.run( ( err, stats ) => {
			// Assert.
			expect( err ).toBe( null );
			expect( stats?.hasErrors() ).toBe( false );
			expect( stats?.hasWarnings() ).toBe( false );

			const fileContent = fs.readFileSync( path.resolve( `/dist/my-app-test.js` ), { encoding: 'utf8' } );

			expect( fileContent ).toMatchSnapshot();

			done();
		} );
	} );
} );
