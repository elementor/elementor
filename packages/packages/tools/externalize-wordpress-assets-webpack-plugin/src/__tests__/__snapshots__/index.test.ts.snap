// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@elementor/externalize-wordpress-assets-webpack-plugin should externalize Elementor & WordPress assets 1`] = `
""use strict";
(self["webpackChunk"] = self["webpackChunk"] || []).push([["my-app-test"],{

/***/ "./.":
/*!***********!*\\
  !*** ./. ***!
  \\***********/
/***/ ((__unused_webpack_module, __webpack_exports__, __webpack_require__) => {

__webpack_require__.r(__webpack_exports__);
/* harmony export */ __webpack_require__.d(__webpack_exports__, {
/* harmony export */   test: () => (/* binding */ test)
/* harmony export */ });
/* harmony import */ var _elementor_editor__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! @elementor/editor */ "@elementor/editor");
/* harmony import */ var _elementor_editor__WEBPACK_IMPORTED_MODULE_0___default = /*#__PURE__*/__webpack_require__.n(_elementor_editor__WEBPACK_IMPORTED_MODULE_0__);
/* harmony import */ var _elementor_editor_panels__WEBPACK_IMPORTED_MODULE_1__ = __webpack_require__(/*! @elementor/editor-panels */ "@elementor/editor-panels");
/* harmony import */ var _elementor_editor_panels__WEBPACK_IMPORTED_MODULE_1___default = /*#__PURE__*/__webpack_require__.n(_elementor_editor_panels__WEBPACK_IMPORTED_MODULE_1__);
/* harmony import */ var _wordpress_element__WEBPACK_IMPORTED_MODULE_2__ = __webpack_require__(/*! @wordpress/element */ "@wordpress/element");
/* harmony import */ var _wordpress_element__WEBPACK_IMPORTED_MODULE_2___default = /*#__PURE__*/__webpack_require__.n(_wordpress_element__WEBPACK_IMPORTED_MODULE_2__);
/* harmony import */ var react__WEBPACK_IMPORTED_MODULE_3__ = __webpack_require__(/*! react */ "react");
/* harmony import */ var react__WEBPACK_IMPORTED_MODULE_3___default = /*#__PURE__*/__webpack_require__.n(react__WEBPACK_IMPORTED_MODULE_3__);
/* harmony import */ var react_dom__WEBPACK_IMPORTED_MODULE_4__ = __webpack_require__(/*! react-dom */ "react-dom");
/* harmony import */ var react_dom__WEBPACK_IMPORTED_MODULE_4___default = /*#__PURE__*/__webpack_require__.n(react_dom__WEBPACK_IMPORTED_MODULE_4__);
/* harmony import */ var _other_package_name__WEBPACK_IMPORTED_MODULE_5__ = __webpack_require__(/*! @other/package-name */ "@other/package-name");
/* harmony import */ var _other_package_name__WEBPACK_IMPORTED_MODULE_5___default = /*#__PURE__*/__webpack_require__.n(_other_package_name__WEBPACK_IMPORTED_MODULE_5__);

			
			
			
			
			
			

			_elementor_editor__WEBPACK_IMPORTED_MODULE_0___default()();
			_elementor_editor_panels__WEBPACK_IMPORTED_MODULE_1___default()();
			_wordpress_element__WEBPACK_IMPORTED_MODULE_2___default()();
			react__WEBPACK_IMPORTED_MODULE_3___default()();
			react_dom__WEBPACK_IMPORTED_MODULE_4___default()();
			_other_package_name__WEBPACK_IMPORTED_MODULE_5___default()();
			
			const test = 'This is test for exporting something';
		

/***/ }),

/***/ "@elementor/editor":
/*!***********************************************!*\\
  !*** external ["elementorPackages","editor"] ***!
  \\***********************************************/
/***/ ((module) => {

module.exports = window["elementorPackages"]["editor"];

/***/ }),

/***/ "@elementor/editor-panels":
/*!*****************************************************!*\\
  !*** external ["elementorPackages","editorPanels"] ***!
  \\*****************************************************/
/***/ ((module) => {

module.exports = window["elementorPackages"]["editorPanels"];

/***/ }),

/***/ "@other/package-name":
/*!*************************************!*\\
  !*** external ["otherPackageName"] ***!
  \\*************************************/
/***/ ((module) => {

module.exports = window["otherPackageName"];

/***/ }),

/***/ "@wordpress/element":
/*!*********************************!*\\
  !*** external ["wp","element"] ***!
  \\*********************************/
/***/ ((module) => {

module.exports = window["wp"]["element"];

/***/ }),

/***/ "react":
/*!**************************!*\\
  !*** external ["React"] ***!
  \\**************************/
/***/ ((module) => {

module.exports = window["React"];

/***/ }),

/***/ "react-dom":
/*!*****************************!*\\
  !*** external ["ReactDOM"] ***!
  \\*****************************/
/***/ ((module) => {

module.exports = window["ReactDOM"];

/***/ })

},
/******/ __webpack_require__ => { // webpackRuntimeModules
/******/ var __webpack_exec__ = (moduleId) => (__webpack_require__(__webpack_require__.s = moduleId))
/******/ var __webpack_exports__ = (__webpack_exec__("./."));
/******/ (window.elementorPackages = window.elementorPackages || {}).myAppTest = __webpack_exports__;
/******/ }
]);"
`;
